server/sonar-server/src/main/java/org/sonar/server/plugins/ws/AvailablePluginsWsAction.java

```diff
@@ -0,0 +1,194 @@
+    controller.createAction("available")
+        .setDescription("Get the list of all the plugins available for installation on the SonarQube instance, sorted by name")
+        .setSince("5.2")
+        .setHandler(this)
+        .setResponseExample(Resources.getResource(this.getClass(), "example-available_plugins.json"));
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    JsonWriter jsonWriter = response.newJsonWriter();
+    jsonWriter.beginObject();
+
+    writePlugins(jsonWriter);
+
+    jsonWriter.close();
+  }
+
+  private void writePlugins(JsonWriter jsonWriter) {
+    jsonWriter.name(ARRAY_PLUGINS);
+    jsonWriter.beginArray();
+    for (PluginUpdate pluginUpdate : retrieveAvailablePlugins()) {
+      writePluginUpdate(jsonWriter, pluginUpdate);
+    }
+    jsonWriter.endArray();
+    jsonWriter.endObject();
+  }
+
+  private void writePluginUpdate(JsonWriter jsonWriter, PluginUpdate pluginUpdate) {
+    jsonWriter.beginObject();
+    Plugin plugin = pluginUpdate.getPlugin();
+
+    jsonWriter.prop(PROPERTY_KEY, plugin.getKey());
+    jsonWriter.prop(PROPERTY_NAME, plugin.getName());
+    jsonWriter.prop(PROPERTY_CATEGORY, plugin.getCategory());
+    jsonWriter.prop(PROPERTY_DESCRIPTION, plugin.getDescription());
+    jsonWriter.prop(PROPERTY_LICENSE, plugin.getLicense());
+
+    writeOrganisation(jsonWriter, plugin);
+
+    writeTermsAndConditions(jsonWriter, plugin);
+
+    writeRelease(jsonWriter, pluginUpdate.getRelease());
+
+    writeUpdate(jsonWriter, pluginUpdate);
+
+    jsonWriter.endObject();
+  }
+
+  private List<PluginUpdate> retrieveAvailablePlugins() {
+    return updateCenterFactory.getUpdateCenter(DO_NOT_FORCE_REFRESH).findAvailablePlugins();
+  }
+
+  private void writeRelease(JsonWriter jsonWriter, Release release) {
+    jsonWriter.name(OBJECT_RELEASE);
+    jsonWriter.beginObject();
+    jsonWriter.prop(PROPERTY_VERSION, release.getVersion().toString());
+    jsonWriter.propDate(PROPERTY_DATE, release.getDate());
+    writeArchive(jsonWriter, release);
+    jsonWriter.endObject();
+  }
+
+  private void writeArchive(JsonWriter jsonWriter, Release release) {
+    jsonWriter.name(OBJECT_ARCHIVE);
+    jsonWriter.beginObject();
+    jsonWriter.prop(PROPERTY_NAME, release.getFilename());
+    jsonWriter.prop(PROPERTY_URL, release.getDownloadUrl());
+    jsonWriter.endObject();
+  }
+
+  private void writeOrganisation(JsonWriter jsonWriter, Plugin plugin) {
+    jsonWriter.name(OBJECT_ORGANIZATION);
+    jsonWriter.beginObject();
+    jsonWriter.prop(PROPERTY_NAME, plugin.getOrganization());
+    jsonWriter.prop(PROPERTY_URL, plugin.getOrganizationUrl());
+    jsonWriter.endObject();
+  }
+
+  private void writeTermsAndConditions(JsonWriter jsonWriter, Plugin plugin) {
+    jsonWriter.name(OBJECT_TERMS_AND_CONDITIONS);
+    jsonWriter.beginObject();
+    jsonWriter.prop(PROPERTY_URL, plugin.getTermsConditionsUrl());
```

Comment:
Please use flat structure as much as possible. In the current case, the field must be "termsAndConditionsUrl".